File: ClockDelegate.m
// ClockDelegate.m
//
// Copyright Michael Schmidt. All rights reserved.


#import "ClockDelegate.h"

static unsigned componentFlags = NSCalendarUnitHour | NSCalendarUnitMinute | NSCalendarUnitSecond;


@implementation ClockDelegate

- (NSDate *)currentDate {

    // Fixed date of iPhone introduction for screenshots:
    // return [NSDate dateWithTimeIntervalSinceReferenceDate:190024881.000000];
    return [NSDate date];
}


- (NSAttributedString *)currentTitleWithAttributes:(NSDictionary *)attributes {

    NSDateComponents *now = [[NSCalendar autoupdatingCurrentCalendar] components:componentFlags fromDate:[self currentDate]];
    NSString *key, *val;

    int hour = (int)now.hour;

    // Some hours have a special name
    if (now.minute == 0) {

        key = [NSString stringWithFormat:@"M%02d", hour % 24];
	    val = _I18N (key);
        
	    if (![val isEqualTo: key])
		    return [[NSAttributedString alloc] initWithString: val attributes: attributes];
    }

    // Does the fuzzy time show the current or next hour name?
    key = [NSString stringWithFormat:@"S%02dh", state % 100];

    if ([_I18N (key) isEqualTo:@"next"]) {

        hour += 1;
    }

    // Format string for current state
    key = [NSString stringWithFormat:@"S%02d", state % 100];
    NSString *format = _I18N (key);
    
    // Hour description
    key = [NSString stringWithFormat:@"H%02d", hour % 12];
    NSString *hour_name = _I18N (key);

    // Build fuzzy time description
    NSString *fuzzy_time = [NSString stringWithFormat: format, hour_name];

    return [[NSAttributedString alloc] initWithString: fuzzy_time
                                           attributes: @{NSFontAttributeName: [NSFont systemFontOfSize: 13]}];
}


- (void)clockTick:(NSTimer *)timer {

    NSDateComponents *now = [[NSCalendar autoupdatingCurrentCalendar] components:componentFlags fromDate:[self currentDate]];

    // Compute the current 30 seconds step of the current hour
    int step = (int)now.minute * 2 + (int)now.second / 30;
    int nextState;
    
    if (step < 2) {

        // ...during the first minute we stick at the full hour
        nextState = 0;
    }
    else if (2 <= step && step <= 5) {

        // ...special state before the first full 5 minutes
        nextState = 1;
    }
    else if (step < 116) {

        // ...rounding to full 5 minute steps
        nextState = 1 + ((step + 4) / 10);
    }
    else {

        // ...round to full next hour
        nextState = 13;
    }

    // Add the current hour to the state
    nextState += now.hour * 100;
    

    // Update if needed
    if (state != nextState) {

        state = nextState;
        clockItem.attributedTitle = [self currentTitleWithAttributes: @{NSFontAttributeName: [NSFont systemFontOfSize: 13]}];
    }
}


- (void)menuNeedsUpdate:(NSMenu *)menu {

    NSDate *now = [self currentDate];
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    
    dateFormatter.dateStyle = NSDateFormatterFullStyle;
    dateFormatter.timeStyle = NSDateFormatterNoStyle;
    dateItem.title = [dateFormatter stringFromDate:now];
    
    dateFormatter.dateStyle = NSDateFormatterNoStyle;
    dateFormatter.timeStyle = NSDateFormatterMediumStyle;
    timeItem.title = [dateFormatter stringFromDate:now];
}


- (void)applicationWillFinishLaunching:(NSNotification *)note {

    // Create status bar item
    NSStatusBar *bar = [NSStatusBar systemStatusBar];
    
    clockItem = [bar statusItemWithLength:NSVariableStatusItemLength];
    
    clockItem.title = @" ";
    [clockItem setHighlightMode:YES];
    clockItem.menu = clockMenu;
    [clockItem setHighlightMode:YES];
    
    state = -1;
    
    // Create a timer object
    clockTimer = [NSTimer scheduledTimerWithTimeInterval:1.0
                                                  target:self
                                                selector:@selector(clockTick:)
                                                userInfo:nil
                                                 repeats:YES];
}

@end

File: ClockDelegate.swift
// ClockDelegate.swift
//
// Copyright Michael Schmidt. All rights reserved.

import Cocoa

@NSApplicationMain
class ClockDelegate: NSObject, NSApplicationDelegate, NSMenuDelegate {

    @IBOutlet weak var clockMenu: NSMenu!
    @IBOutlet weak var dateItem: NSMenuItem!
    @IBOutlet weak var timeItem: NSMenuItem!

    var clockItem: NSStatusItem?
    var clockTimer: Timer?

    var state: Int = -1 {
        didSet {
            clockItem?.attributedTitle = currentAttributedTitle()
        }
    }

    // Reads localized format strings for fuzzy time formatting
    func localizedString(format: String, value: Int) -> String? {

        let key = String(format: format, value)
        let localized = NSLocalizedString(key, comment: "")
        guard (localized != key) else { return nil }

        return localized
    }

    // Creates an attributed title string representing the current time as a fuzzy description
    func currentAttributedTitle() -> NSAttributedString {

        let now = Calendar.autoupdatingCurrent.dateComponents([.hour, .minute, .second], from: NSDate() as Date)
        let attributes = [ NSFontAttributeName: NSFont.systemFont(ofSize: 13) ]

        // Some special hours may have a predefined title string like e.g. "midnight"
        if now.minute == 0 {
            if let title = localizedString(format: "M%02d", value:now.hour! % 24) {
                return NSAttributedString.init(string: title, attributes: attributes)
            }
        }

        // Time descriptions may refer to the current or the following hour
        let hourOffset : Int

        if localizedString(format: "S%02dh", value:state % 100) == "next" {
            hourOffset = 1
        }
        else {
            hourOffset = 0
        }

        // Build the fuzzy time description
        let title: String

        if let format = localizedString(format: "S%02d", value: state % 100),
           let hourName = localizedString(format: "H%02d", value: (now.hour! + hourOffset) % 12) {
            title = String(format: format, hourName)
        }
        else {
            title = "Error"
        }

        return NSAttributedString.init(string: title, attributes: attributes)
    }

    func clockTick() {

        let now = Calendar.autoupdatingCurrent.dateComponents([.hour, .minute, .second], from: NSDate() as Date)

        // Compute the current 30 seconds step of the current hour
        let step = now.minute! * 2 + now.second! / 30

        // Next state of internal clock, initialized with information about current hour of day
        var nextState = now.hour! * 100;

        if (step < 2) {

            // ...during the first minute we stick at the full hour
            nextState = 0;
        }
        else if (2 <= step && step <= 5) {

            // ...special state before the first full 5 minutes
            nextState = 1;
        }
        else if (step < 116) {

            // ...rounding to full 5 minute steps
            nextState = 1 + ((step + 4) / 10);
        }
        else {

            // ...round to full next hour
            nextState = 13;
        }

        // Update only if needed
        if state != nextState {
            state = nextState
        }
    }

    func menuNeedsUpdate(_ menu: NSMenu) {

        // Update menu contents with current date and time
        let now = NSDate() as Date
        let dateFormatter = DateFormatter()

        dateFormatter.dateStyle = .full
        dateFormatter.timeStyle = .none
        dateItem.title = dateFormatter.string(from: now)

        dateFormatter.dateStyle = .none
        dateFormatter.timeStyle = .medium
        timeItem.title = dateFormatter.string(from: now)
    }

    func applicationWillFinishLaunching(_ aNotification: Notification) {

        // Create the status bar item
        let bar = NSStatusBar.system()
        clockItem = bar.statusItem(withLength: NSVariableStatusItemLength)
        clockItem?.title = " "
        clockItem?.menu = clockMenu
        clockItem?.highlightMode = true
        clockItem?.autosaveName = "me.mschmidt.FuzzyClock"
        clockItem?.behavior = [ .removalAllowed, .terminationOnRemoval ]

        // Install timer for clock updates
        clockTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { _ in self.clockTick() }
    }
}
